15 gradedm

dkosmas opened the pull request:
Updated randomGeneratingSet with list input so that a list of integers can be inputted, for a graded model for set number of monomials. Also includes an error if the list is not either all integers or all real numbers (do we want to allow for a mixed model?).

dkosmas committed:
Update RandomMonomialIdeals.m2

dkosmas committed:
Update RandomMonomialIdeals.m2

dkosmas committed:
Update RandomMonomialIdeals.m2

dkosmas committed:
Update RandomMonomialIdeals.m2

dkosmas committed:
Update RandomMonomialIdeals.m2

dkosmas committed:
Update RandomMonomialIdeals.m2

dkosmas committed:
Update RandomMonomialIdeals.m2

dkosmas committed:
Update RandomMonomialIdeals.m2

dkosmas committed:
Update RandomMonomialIdeals.m2

dkosmas committed:
Update RandomMonomialIdeals.m2

dkosmas committed:
Update RandomMonomialIdeals.m2

dkosmas committed:
Update RandomMonomialIdeals.m2

dkosmas committed:
Update RandomMonomialIdeals.m2

dkosmas committed:
Update RandomMonomialIdeals.m2

dkosmas committed:
Update RandomMonomialIdeals.m2

dkosmas committed:
Update RandomMonomialIdeals.m2

dkosmas committed:
Update RandomMonomialIdeals.m2

sonjapetrovic committed:
small doc changes

- slightly changed formatting for inputs when list is an option
- changed M={1,1,1} to M={2,1,1} just for fun (so it’s not only 1).

sonjapetrovic committed:
added SeeAlso between randomGeneratingSet and randomGeneratingSets

sonjapetrovic committed:
edited documentation again

- fixed a SeeAlso error I had
- changed inputs to randomGeneratingSets also (the plural version) to
reflect gradedM case
- cut doc node text for plural and simply had it refer to singular (we
were gonna do that anyway, just forgot)

Sondzus commented:
Great! 
I added some small documentation updates @dkosmas ; see if you see any additional issues. 

Someone else should also check we are not missing anything in the graded version of fixed M -- perhaps let's let see if @sdgpfo  or @dwilburne or someone else comment on it, before we try to merge the pull request.

dkosmas commented:
Just read over the changes, I'm good with all of them. If someone else could review them, that would be great.

sdgpfo commented:
Great thanks @dkosmas and @Sondzus. It all looks good to me. 

Good question regarding mixed case, @dkosmas, my first instinct is not to allow that, because I don't see a reason why we would need to or want to use that.

My only minor suggestions for now are: 
1. to rename p to pOrM or something similar to indicate that it serves a dual role depending on the input type. (I suppose there is the risk that a 1.0-value will be entered for p as an integer and cause an error to be thrown, but that seems to be a general issue, and I'm not sure how to mitigate it for the time being.) 
2. for the minimal strategy error message to be changed to add the word "yet", as in "Minimal not implemented yet for fixed size ER mode."
3. for the third assert statement in the first test to be removed, namely:
    assert (set L#1===set L#2) 
as this is already checked implicitly by the combination of the prior two tests.

Sondzus commented:
@sdgpfo  you mean "pOrM" in the code only, right? The user should not get confused: the user sees that he/she can put in either a list of ps (real numbers) or list of Ms (integers) and both are documented.

sonjapetrovic committed:
added word in documentation

sdgpfo commented:
@Sondzus <https://github.com/Sondzus> sorry not following your question: I
meant  for the specific method dealing with a list input of M's or p's. I
have made the changes on my laptop but didnt want to push, until you guys
agreed with them. This doesn't change the functionality it just makes it
obvious to the coder what the input parameter is. Sorry, I hope I answered
your question.

Sondzus commented:
@sdgpfo  OK go ahead and push your changes if all you are doing is chaing the code part of dealing with the List - pOrM - is fine with both of us! :) 

(But if you are not on your laptop I can make the change too.)

dkosmas committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into 15-gradedM

# Conflicts:
#	RandomMonomialIdeals.m2

dkosmas committed:
Create RandomMonomialIdeals.m2

dkosmas committed:
Create RandomMonomialIdeals.m2

sdgpfo commented:
@Sondzus <https://github.com/Sondzus> Done but my Internet is slow, still
uploading.

On Jun 8, 2017 4:46 PM, "SP" <notifications@github.com> wrote:

> @sdgpfo <https://github.com/sdgpfo> OK go ahead and push your changes if
> all you are doing is chaing the code part of dealing with the List - pOrM -
> is fine with both of us! :)
>
> (But if you are not on your laptop I can make the change too.)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/RandCommAlg/RMI/pull/23#issuecomment-307126012>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ARvaODjRO8gJ7txZXIWpAtwVpsYFbypLks5sCAkxgaJpZM4NxjZu>
> .
>

dkosmas committed:
Create RandomMonomialIdeals.m2

dkosmas committed:
Create RandomMonomialIdeals.m2

dkosmas committed:
Create RandomMonomialIdeals.m2

dkosmas committed:
Create RandomMonomialIdeals.m2

dkosmas committed:
Create RandomMonomialIdeals.m2

sdgpfo committed:
minor changes

renamed variable p to pOrM in graded version of randomMonomialSet that accepts a list, for clarity: commented out redundant test-propose delete; added some minor language in documentation.

Sondzus commented:
thanks @sdgpfo  I see your changes now. I'm fine with this.
@dkosmas are you good with these too? 

If so, I (or you or someone else)  can go ahead and merge this pull request because this was the last pending item we were waiting on.

sdgpfo commented:
@Sondzus @dkosmas Thanks: my commit took too long due to internet issues, and by the time I got to wifi, you had progressed with your changes and I had to resolve some conflicts. Nonetheless, if you think of a better looking name than pOrM please go ahead.

sdgpfo committed:
minor language change

forgotten from prior commit+conflict resolution

dkosmas commented:
I think the name pOrM is more fitting. Should be update the documentation to reflect this? As of now, it has that inputs are either p, a real number or a list of real numbers, or M, an integer or list of integers. Should we split that to p, M, or pOrM?

Sondzus commented:
@dkosmas  I would vote for *not* changing the documentation to pOrM. The documentation is pretty clear as is, in my opinion.

dkosmas commented:
Ok, I'm fine with that. I wanted to bring it up, just in case we did want to change the documentation. I'm good with all changes then.

Sondzus commented:
Absolutely - thanks for bringing it up - it is better to discuss all options as we go - this is how improvements happen! :)

sdgpfo commented:
@dkosmas @Sondzus 
Thanks for asking the question @dkosmas.

Currently this is what is written for input for p:
(randomMonomialIdeals) 
p, a real number, or a list , probability to select a monomial
(randomMonomialSet) & (randomMonomialSets)
p, a real number, the probability of selecting a monomial, or a list of real numbers whose i-th entry is the probability of selecing a monomial of degree i

I think we should change the documentation to reflect the actual inputs, namely:

(randomMonomialIdeals) 
p, a real number, the probability of selecting each monomial
pOrM, a list controlling the number of monomials in the generating monomial set of each degree for the graded ER model; specifically this input is either a list of real numbers between 0 and 1 inclusive, whose i-th entry is the probability of including a monomial of degree i in the generating set, or it is a list of nonnegative integers whose i-th entry is the number of monomials of each degree to include in the generating set; if this number is larger than the maximum possible, the latter is used instead.

(randomMonomialSet)  & (randomMonomialSets)
p, a real number, the probability of selecting each monomial
pOrM, a list controlling the number of monomials in the generating set of each degree for the graded ER model; specifically this input is either a list of real numbers between 0 and 1 inclusive, whose i-th entry is the probability of including a monomial of degree i in the monomial set, or it is a list of nonnegative integers whose i-th entry is the number of monomials of each degree to include in the monomial set; if this number is larger than the maximum possible, the latter is used instead.

Sondzus commented:
@sdgpfo I think your list of what we currently have is incomplete; here is what we currently have in the documentation:

(randomMonomialIdeals)
p, a real number, or a list , probability to select a monomial
M, an integer, maximum number of monomials in each generating set for the ideal [*this shoudl change to allow for graded, I don't see the change in my current version*]

(randomMonomialSet) & (randomMonomialSets)
p, a real number, the probability of selecting a monomial, or a list of real numbers whose i-th entry is the probability of selecing a monomial of degree i
M, an integer, number of monomials in the set, or a list of integers whose i-th entry is the number of monomials of degree i in each set

Since we list both p and M as options I am voting against using pOrM (which does not look so good at all in the documentation and having p being real or list of reals and M being integer or a list of integers is much more natural to me).

Sondzus commented:
I want to update my comment above:
We also are required to document the overloaded versions
randomMonomialSet(ZZ,ZZ,RR)
randomMonomialSet(ZZ,ZZ,ZZ)
separately, so keeping p and M separate is the most natural thing, as far as I'm concerned.

Here is an example of a certified package and how they document the overloaded method:
[http://www.math.uiuc.edu/Macaulay2/doc/Macaulay2-1.9.2/share/doc/Macaulay2/EdgeIdeals/html/_complement__Graph.html]
We are currently following similar guidelines for formatting our documentation. @sdgpfo

Sondzus commented:
Here is an example of a certified package and how they document the overloaded method:

[http://www.math.uiuc.edu/Macaulay2/doc/Macaulay2-1.9.2/share/doc/Macaulay2/EdgeIdeals/html/_complement__Graph.html]
We are currently following similar guidelines for formatting our documentation. @sdgpfo

sdgpfo commented:
Thanks @Sondzus. I understand what you're saying and I definitely agree that we should respect documenting overloaded functions separately. p and M should be kept separate where they are used separately, but they are not used separately in the graded version of the method (hence the name pOrM). The appropriate p and M and pOrM descriptions should be copy-pasted in the documentation of the appropriate respective version of the function, namely:
 randomMonomialSet(ZZ,ZZ,RR)
p-
randomMonomialSet(ZZ,ZZ,ZZ)
M-
 randomMonomialSet(ZZ,ZZ,List)
pOrM -

and the same for other overloaded versions of randomMonomialSets and randomMonomialIdeals.

(By the way, as far as I can tell, we don't currently document overloaded methods separately, but if that is the convention we should of course)

Sondzus commented:
Oh, I see what you are saying now @sdgpfo , thanks.  

Are there any better names for pOrM ? Just L for a list? 

(Sorry if I wasn't clear: by 'separately' i didn't mean separate documentation nodes but separate lines, i.e., we do not write (ZZ,ZZ,ZZ or RR) . )

Sondzus commented:
Can we just say this way:

randomMonomialSet(ZZ,ZZ,List)
L: List of real numbers whose i-th entry is the probability of selecting a monomial of degree i, or a list of integers whose i-th entry is the number of monomials of degree i in each set.

and keep pOrM in the code (instead of just "L").

sdgpfo commented:
@Sondzus I'm not crazy about the name either but I can't think of a better name that contains all three of "p", "or", and "M," which I think is important. 
Is p.or.M better? 
p.or.M.graded? 
p.or.M.list?
(really these methods should have been overloaded versions of the same, but I imagine the problem was that there was no way to distinguish between a list of RRs and a list of ZZs.)

sonjapetrovic committed:
here are my proposed changes for documentation for pOrM

I just now see Despina’s latest comment and I’m not crazy about any of
the names.
Maybe we can come up with a better name for “L” later - but note it
only appears when we document what input types are. Maybe

Sondzus commented:
@sdgpfo @dkosmas I just pushed Despina's proposed changes.
But I put "L" for a list. It only appears in 1 place.
*edit*: no need for using input name for a list! check the latest change see if you like it. 

@dkosmas  I did *not* update randomMonomialIdeals documentation accordingly - so as to avoid doing double work if we decide we don't like what just happened.

sonjapetrovic committed:
sorry I pushed a typo! fixed.

sonjapetrovic committed:
check it out: we don't need a name for the list!!

how about this? :)

Sondzus commented:
@sdgpfo in case your internet is slow:

randomMonomialSet -- randomly generates a list of monomials, up to a given degree

Synopsis

Usage:
randomMonomialSet(ZZ,ZZ,RR)
randomMonomialSet(ZZ,ZZ,ZZ)
randomMonomialSet(ZZ,ZZ,List)
Inputs:
 n, an integer, number of variables
 D, an integer, maximum degree
 p, a real number, the probability of selecting a monomial,
 M, an integer, number of monomials in the set,
 a list, of real numbers whose i-th entry is the probability of selecting a monomial of degree i, or of integers whose i-th entry is the number of monomials of degree i in each set

sonjapetrovic committed:
minor typo fix

 (selecing->selecting)

Sondzus commented:
@sdgpfo  I want to put in your text into the List:
"
pOrM, a list controlling the number of monomials in the generating set of each degree for the graded ER model; specifically this input is either a list of real numbers between 0 and 1 inclusive, whose i-th entry is the probability of including a monomial of degree i in the monomial set, or it is a list of nonnegative integers whose i-th entry is the number of monomials of each degree to include in the monomial set; if this number is larger than the maximum possible, the latter is used instead.
"

However, I don't understand "the latter is used instead". ? Can you please explain that to me? Sorry.

Sondzus commented:
@sdgpfo  --- NEVER MIND my last question  --- I understand. I think that comment is better looking down in the text (we already have it). ?

sdgpfo commented:
@Sondzus @dkosmas 
Thanks @Sondzus. I agree with your re-write of the description of porM (or L as you have it written), it's much shorter and still precise. However, I do not agree with giving a name to the input parameter in the documentation that is different in the code: there is a reason why the name of the parameter is included in the documentation to begin with.

sonjapetrovic committed:
edited List description using Despina's text

sdgpfo commented:
Oh, my internet is slow, sorry our conversation is out of sync.

So I am not sure now have we come to an agreement :-)

Sondzus commented:
@sdgpfo 
So apparently there is also a reason why you don't have to put the name of the input parameter in the documentation at all (!).

See my latest change I just pushed using your text. Do you prefer your (long) text or my (short) text I had a minute ago? 

randomMonomialSet -- randomly generates a list of monomials, up to a given degree

Synopsis

Usage:
randomMonomialSet(ZZ,ZZ,RR)
randomMonomialSet(ZZ,ZZ,ZZ)
randomMonomialSet(ZZ,ZZ,List)
Inputs:
n, an integer, number of variables
D, an integer, maximum degree
p, a real number, the probability of selecting a monomial,
M, an integer, number of monomials in the set,
a list, controlling the number of monomials in the generating set of each degree for the graded ER model. ; s Specifically, this input is either a list of real numbers between 0 and 1, inclusive, whose i-th entry is the probability of including a monomial of degree i in the monomial set, or it is a list of nonnegative integers whose i-th entry is the number of monomials of each degree to include in the monomial set; if this number is larger than the maximum possible number of monomials, the method returns all such monomials.

Sondzus commented:
@sdgpfo  I personally prefer my shorter version of the List description, but if you want your longer version in the documentation I'd be fine leaving it that way.

sonjapetrovic committed:
2 typos fixed

dkosmas commented:
I prefer the shorter version as well. Whatever is agreed upon will also need to be updated in randomMonomialSets and randomMonomialIdeals, both in code and documentation. 
In terms of functionality, randomMonomialIdeals should already be able to run the graded model, since it calls randomMonomialSets (I'm not sure if this is what you meant by this "shoudl change to allow for graded, I don't see the change in my current version", @Sondzus)

Sondzus commented:
Thanks @dkosmas . I am working on some more edits to try to unify everything. I propose moving @sdgpfo 's long text into the text of the documentation nodes, to further explain what the list does, and keeping the short list description in Input type explanation. 
Let me push the change so you can both see what it looks like and see if it makes sense / is consistent. :)

Sondzus commented:
@dkosmas  my weird comment about changing to allow for graded made no sense, my apologies:
I simply noticed we can use randomMonomialIdeals( ..., List, ..) and call the graded versions of randomMonomialSets; but the documentation for randomMonomialIdeals only specified a list of ps and not a list of Ms. 
But I had *forgotten* that we don't use Minimal for graded M!

sdgpfo commented:
@Sondzus @dkosmas Hope that this message is now in sync: I also like @Sondzus 's shorter description better, but perhaps we should keep the longer text in comments in the description part after the input as it includes information we need to give, but can be improved/clarified/shortened before added.

sonjapetrovic committed:
fixed documentation as per discussion.

Sondzus commented:
@dkosmas check new push, please, for consistency.:)
@sdgpfo  great - we have come to an agreement - I think - according to our out-of-sync comments. I just pushed the changes and they are like you said in your latest comment (short description + kept your longer text in the explanation below where we show examples. it is more thorough and very informative to the user so I like it this way!) :)

Sondzus commented:
@sdgpfo  here is what it looks like now:
.....
randomMonomialSet -- randomly generates a list of monomials in fixed number of variables up to a given degree

Synopsis

Usage:
randomMonomialSet(ZZ,ZZ,RR)
randomMonomialSet(ZZ,ZZ,ZZ)
randomMonomialSet(ZZ,ZZ,List)
Inputs:
n, an integer, number of variables
D, an integer, maximum degree
p, a real number, the probability of selecting a monomial,
M, an integer, number of monomials in the set,
a list, of real numbers whose i-th entry is the probability of selecing a monomial of degree i, or of integers whose i-th entry is the number of monomials of degree i in each set
.....
[some more text ]
.....
The input of type List controls the number of monomials in the generating set of each degree for the graded ER model. Specifically, this input is either a list of real numbers between 0 and 1, inclusive, whose i-th entry is the probability of including a monomial of degree i in the monomial set, or it is a list of nonnegative integers whose i-th entry is the number of monomials of each degree to include in the monomial set. Consider the following two examples: 
.....
[and then the 2 graded examples we had, one for ps one for Ms.]

dkosmas commented:
In the function, randomMonomialSets and randomMonomialIdeals need to have the name of the inputs updated from p to pOrM. Also, in the documentation, randomMonomialIdeals doesn't have the updated inputs and text and examples for the list input. I'll update the first thing I mentioned, but save the later for when we confirm what is now in the documentation is agreed upon.

dkosmas committed:
Create RandomMonomialIdeals.m2

Sondzus commented:
@dkosmas 
I agree with pOrM change to randomMonomialSets.

BUT:
Are you sure the List in randomMonomialIdeals can be ps or Ms? 
Regarding documentation: I didn't update it the same as for randomMonomialSet on purpose. Namely, it doesn't seem to me that  we can use the same List description text for randomMonomialIdeals as we are using for randomMonomialSet: we can only do the first part (list of ps).

If I try to do graded M for randomMonomialideal, this is what I get:
...
i12 :    randomMonomialIdeals(2,3,{2,1,1},1)
stdio:12:4:(3): error: Minimal not implemented for fixed size ER model
...
Am I doing something wrong?

dkosmas commented:
@Sondzus Oops, I just noticed that. randomMonomialIdeals automatically uses "Minimal" strategy, but randomMonomialSet gives an error if Strategy=>"Minimal". Should be an easy fix, I'll take care of that quickly and let you know when I get it working.

Sondzus commented:
also @dkosmas  when you pass a list of options you can just pass "o" and not the list of entries of "o"s - I see the conflict so let's respect the master file in this.

Sondzus commented:
@dkosmas  Hmmmm... I don't think you can promise to construct an ideal with M_i minimal generators of degree i. That's why the error was there in the first place. So we should just promise to construct an ideal with M_i *generators* (they need not be minimal). Just saying this so we are careful in the documentation of it. :)

dkosmas commented:
@Sondzus, I definitely agree with that statement. Until we develop a minimal strategy for the fixed M model, then we cannot guarantee M_i minimal generators. It is, as you put it, M_i generators, or at most M_i minimal generators.

dkosmas committed:
updated to allow graded fixed M model

dkosmas committed:
Update RandomMonomialIdeals.m2

dkosmas commented:
@Sondzus @sdgpfo randomMonomialIdeals now works with graded fixed M model. My bad for not catching that bug sooner!

Sondzus commented:
@dkosmas  
Re: conflicts with master, the line you fixed should be
    apply(N,i-> randomMonomialSet(n,D,pOrM,o))
and not 
    apply(N,i-> randomMonomialSet(n,D,pOrM,
	                            Coefficients=>o.Coefficients,
				    VariableName=>o.VariableName,
				    Strategy=>o.Strategy))
because the former is the streamlined approach, applied across the package in the master version. :) 
But since you're working on it I don't want ot make changes or resolve this conflict while you're in the file.

Sondzus commented:
I get some errors trying to installPackage with the latest version. I don't understand what's happening. 

i13 : randomMonomialIdeals(2,3,{0.0, 1.0, 1.0},1)
stdio:7:1:(3): error: no method found for applying idealsFromGeneratingSets to:
     argument   :  null (of class Nothing)

dkosmas commented:
@Sondzus I did not know that you could just use o, I will definitely use that from now on.
As for the error, it seems that fixing it for graded fixed M broke it for graded p, and I missed that since I did load instead of install. Give me some time to look at it, I'll see how I can fix it.

dkosmas committed:
potentially fixed

dkosmas commented:
Ok, I think I fixed it. It installed and there were no errors when I did check "RandomMonomialIdeals", so it should be good now. What it was doing was, since it was written as "B:=" twice, I think because if all(pOrM, q->instance(q,RR)) was first, the else if statement was overwriting the B generated by the original statement. @Sondzus, please confirm that the function now works for you.

Sondzus commented:
@dkosmas thanks, package now installs without issue. But I have a question: 

1. Looking at examples (running line-by-line), I notice the following:
i17 :    randomMonomialIdeals(2,3,{0.0, 1.0, 1.0},1)
o17 = {monomialIdeal (x_1^2 x_1x_2 x_2^2)}
--- how come there are no monomials of degree 3? isn't p_1=0, p_2=1, p_3=1?  Or am I reading this wrong again... 

2. 
i24 : randomMonomialIdeals(2,3,{1,1,1},1)
--- returns x_1, x_2^2
randomMonomialIdeals(2,3,{1,2,1},1)
--- returns x_1^2, x_2. 
I expected to see 2 (not 1) monomials of degree 2 in the latter, and 1 of degree 3 in both of these. 

(We need to add documentation for gradedM now, but of course you knew that already, first we should make sure the method works.)

dkosmas commented:
Correct me if I'm wrong, but doesn't the monomialIdeal object not list any unnecessary generators? That would clearly answer question 1 (since all of the degree 2 monomials are included in the ideal, so all degree 3 monomials are clearly multiples), and give a reasonable explanation for question 2.

Sondzus commented:
You are not wrong @dkosmas ! I am wrong! Sorry about that. :)  Clearly: 
i26 : monomialIdeal (x_1,x_1*x_2) 
o26 = monomialIdeal x_1

But let me tell you why I am wrong:
while the monomialIdeal object clears multiples, the ideal object does not (the computation may be expensive in general):
i27 : ideal(x_1,x_1*x_2)
o27 = ideal (x_1, x_1 x_2 )
i28 : mingens oo
o28 = | x_1 |

So, my bad - I was too used to ideal and not monomialIdeal. 

I think we should add a Caveat to the documentation: that monomialIdeal displays only minimal generators so p={1.0,1.0} will not include anyone of degree 2 because they are not minimal. "Caveat" is a key word just like "Description" or "SeeAlso".

Sondzus committed:
Merge branch 'master' into 15-gradedM

dkosmas committed:
Updated documentation to randomMonomialIdeals

dkosmas committed:
Update to randomMonomialIdeals documentation

dkosmas committed:
Create RandomMonomialIdeals.m2

dkosmas committed:
Update RandomMonomialIdeals.m2

dkosmas committed:
Update to randomMonomialIdeals examples

dkosmas commented:
I've updated randomMonomialIdeals documentation to reflect that it can handle the input of a list of integers, as well as added a Caveat section. Can you read it over @Sondzus?

Sondzus committed:
Minor changes to documentation

- also removed 2 unnecessary files from branch
- removed 2 exported symbols for file writing that we don’t need any
more.

Sondzus commented:
Great!! Thanks for making these changes @dkosmas . I made a small edit just now; see if you approve.

dkosmas commented:
Looks good to me! Anything else we want to do with it before merging?

Sondzus commented:
I think we are done with this - but of course if someone comes back later (e.g. Dane, Despina) and wants to edit more, they can do that. I think we can merge it now since the skeleton is decided on and the content is pretty-much done. :)

Sondzus commented:
@dwilburne let us know if you plan to look this over and make any (even minor) edits; if not, you can - or I will - go ahead and merge this pull request.

dwilburne commented:
@Sondzus I have nothing to add or change at this time, so I am merging.
